plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -0,0 +1,240 @@
+import org.elasticsearch.threadpool.ThreadPool;
+
+/**
+ * Abstract base for scrolling across a search and executing bulk indexes on all
+ * results.
+ */
+public abstract class AbstractAsyncBulkIndexByScrollAction<Request extends AbstractBulkIndexByScrollRequest<Request>, Response extends BulkIndexByScrollResponse>
+        extends AbstractAsyncBulkByScrollAction<Request, Response> {
+
+    private final AtomicLong noops = new AtomicLong(0);
+    private final ScriptService scriptService;
+    private final CompiledScript script;
+
+    public AbstractAsyncBulkIndexByScrollAction(ESLogger logger, ScriptService scriptService, Client client, ThreadPool threadPool,
+            Request mainRequest, SearchRequest firstSearchRequest, ActionListener<Response> listener) {
+        super(logger, client, threadPool, mainRequest, firstSearchRequest, listener);
+        this.scriptService = scriptService;
+        if (mainRequest.script() == null) {
+            script = null;
+        } else {
+            script = scriptService.compile(mainRequest.script(), ScriptContext.Standard.UPDATE, mainRequest);
+        }
+    }
+
+    /**
+     * Build the IndexRequest for a single search hit. This shouldn't handle
+     * metadata or the script. That will be handled by copyMetadata and
+     * applyScript functions that can be overridden.
+     */
+    protected abstract IndexRequest buildIndexRequest(SearchHit doc);
+
+    /**
+     * The number of noops (skipped bulk items) as part of this request.
+     */
+    public long noops() {
+        return noops.get();
+    }
+
+    @Override
+    protected BulkRequest buildBulk(Iterable<SearchHit> docs) {
+        BulkRequest bulkRequest = new BulkRequest(mainRequest);
+        ExecutableScript executableScript = null;
+        Map<String, Object> scriptCtx = null;
+
+        for (SearchHit doc : docs) {
+            IndexRequest index = buildIndexRequest(doc);
+            copyMetadata(index, doc);
+            if (script != null) {
+                if (executableScript == null) {
+                    executableScript = scriptService.executable(script, mainRequest.script().getParams());
+                    scriptCtx = new HashMap<>(16);
+                }
+                if (false == applyScript(index, doc, executableScript, scriptCtx)) {
+                    continue;
+                }
+            }
+            bulkRequest.add(index);
+        }
+
+        return bulkRequest;
+    }
+
+    /**
+     * Copies the metadata from a hit to the index request.
+     */
+    protected void copyMetadata(IndexRequest index, SearchHit doc) {
+        index.parent(fieldValue(doc, ParentFieldMapper.NAME));
+        copyRouting(index, doc);
+        // Comes back as a Long but needs to be a string
+        Long timestamp = fieldValue(doc, TimestampFieldMapper.NAME);
+        if (timestamp != null) {
+            index.timestamp(timestamp.toString());
+        }
+        index.ttl(fieldValue(doc, TTLFieldMapper.NAME));
```

Comment:
I thought about being fancy with these like I am with routing but I decided against it because they are both deprecated features. I'll add some validation that the destination doesn't have either the ttl or timestamp.